ddtrace/_trace/utils_botocore/span_pointers/dynamodb.py

```diff
@@ -154,10 +155,11 @@ def _extract_span_pointers_for_dynamodb_putitem_response(
         ]
 
     except Exception as e:
```

Comment:
It's true that avoiding exceptions-as-control-flow is generally advisable in Python. In this case I'd first look to shrink the `try` block as much as possible and scope the Exception type to only the specific types that are expected from the calls in the `try` block. Certainly if it's possible to remove the source of the exception at a `raise` call that would be preferable.

Reply:
[_aws_dynamodb_item_primary_key_from_item](https://github.com/DataDog/dd-trace-py/blob/9e7acd09ba1d39f92761fc52e68677390b2d93c6/ddtrace/_trace/utils_botocore/span_pointers/dynamodb.py#L279) seems like it could be fairly easily changed to return `Optional[_DynamoDBItemPrimaryKey]` (ie. `_DynamoDBItemPrimaryKey | None`) and we could remove this [raise](https://github.com/DataDog/dd-trace-py/blob/9e7acd09ba1d39f92761fc52e68677390b2d93c6/ddtrace/_trace/utils_botocore/span_pointers/dynamodb.py#L284) in favor of `log.debug(...)` and `return None`

Reply:
i was trying to keep the try blocks pretty large to make sure that we *don't* crash the customers' apps. i can refactor this to use optional returns instead of exceptions for flow control, but we'd need to update the python lambda layer first to make sure that it behaves correctly with this api change. i'll make a separate ticket to do this.

Reply:
is it actually better to replace try-excepts and occasional raise-Exceptions with a lot of extra "return if none" checks? can we get meaningful performance metrics for this somehow? https://datadoghq.atlassian.net/browse/SVLS-5975

Reply:
i do also think that the exceptions here are actually pretty exceptional and that we don't expect to actually see them in the normal flow of this code. though, maybe we can explore that in the followup pull request for that new ticket.

Reply:
> i do also think that the exceptions here are actually pretty exceptional and that we don't expect to actually see them in the normal flow of this code
The piece to be very careful about is whether it's going to occur frequently for some users of the library. For example, if a user always has multiple primary keys and calls DDB 50k/s, they're going to feel the pain from this. Although that'd be a small subset of users, it could make or break a performance analysis of us vs another APM vendor

> can we get meaningful performance metrics for this somehow?
To get this, we'd first need to create [microbenchmarks](https://benchmarking.us1.prod.dog/trends?timeStart=1678871996072&timeEnd=1681463996072&projectId=3) for this particular case. From there, you can evaluate if perf got better or worse. If you want to do something more lightweight/less repeatable, you can run benchmarks locally using a tool of your choice

Reply:
The doc on benchmarks is found here and there should be plenty of examples: https://github.com/DataDog/dd-trace-py/blob/main/benchmarks/README.rst Although I think it is fair to make perf testing a separate ticket.

Reply:
I'm also not sure this is something that needs its own full benchmark test. If you could do some local testing of exceptions vs multiple "return if none" checks and found little difference over a lot of runs, then it's probably fine to stick with the current approach and that info should just be added to the description.

Reply:
i've made the refactors to remove all the `raise *` calls in this code that i could fine anyway